### PR TITLE
Feat/issue 154/initial classification pipeline

### PR DESCRIPTION
**Pull request by TicaGit**
Solves issue #154 and partially addresses #155.

This PR introduces a pipeline for classifying the soil description of layers into USCS classes.

For testing the classification workflow, a dummy classifier can be used. Additionally, I have started implementing the Baseline classifier, which performs keyword matching between the USCS class descriptions and the layer's material description.

The description in the README.predictions-json.md was also updated to match the current format.

### CODE REVIEW

File: src/description_classification/main.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    logger.info(f"classification metrics: {classification_metrics.to_json()}")
+    logger.debug(f"classification metrics per class: {classification_metrics.to_json_per_class()}")
+
+    if mlflow_tracking:
+        mlflow.end_run()
+
+
+if __name__ == "__main__":
+    # launch with: python -m src.description_classification.main -f data/geoquat_ground_truth.json
+    # or with: boreholes-classify-descriptions  -f data/geoquat_ground_truth.json
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/main.py:99]: Do you think it would make sense to add an option to only run the classification on description from files that are in a specific dataset (train/test/validation). Once, we have a proper classifier, always running the classification on the full Geoquat data might become quite slow. Also, the split would allow us to check whether we might be overfitting.

**Review by stijnvermeeren-swisstopo — CHANGES_REQUESTED** [review]
Is there a way (e.g. JSON file / MLFLow artifact?) to see the actual predictions for every description?

File: src/description_classification/main.py (hunk 1 above)
Comment (TicaGit) [reply src/description_classification/main.py:99]: Yes, I think in the future we want to be able to pass the predictions.json file generated. by the extraction pipeline. This requires to map each of the detected layers in the predictions.json to the real layers in the ground truth, so it might require more work.

Comment (TicaGit) [reply src/description_classification/main.py:99]: Regarding your question about outputting the actual predictions for every description, no there is no such option yet. But on MLFLOW, you can see the detail for each predicted class.

File: src/description_classification/evaluation/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    classes = set(predictions) | set(ground_truth)  # Get all unique classes
+    pred_counter = Counter(predictions)
+    gt_counter = Counter(ground_truth)
+
+    metrics_per_class = {}
+    for cls in classes:
+        tp = min(pred_counter[cls], gt_counter[cls])  # Intersection
+        fp = pred_counter[cls] - tp
+        fn = gt_counter[cls] - tp
+        metrics_per_class[cls] = Metrics(tp=tp, fp=fp, fn=fn)
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/evaluation/evaluate.py:197]: I have some doubts about this computation. Does this even check if the predictions are aligned with the ground truth? If we randomly shuffle the correct classes, wouldn't this still give us a 100% F1 score, even though most descriptions will have an incorrect class?

I would have expected the computation to go more along these lines (ideally we would also avoid the `zip`, and instead already pass each prediction and corresponding ground truth value pairwise together to this method):
```
        tp = sum([1 for pred, gt in zip(predictions, ground_truth) if pred == gt == cls])
        fp = sum([1 for pred, gt in zip(predictions, ground_truth) if pred == cls and gt != cls])
        fn = sum([1 for pred, gt in zip(predictions, ground_truth) if pred != cls and gt == cls])
```

Comment (TicaGit) [reply src/description_classification/evaluation/evaluate.py]: Yes, you're totally right, great spot!

File: src/description_classification/utils/data_loader.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                uscs_class = map_most_similar_uscs(layer["uscs_1"])
+                layer_descriptions.append(
+                    LayerInformations(
+                        filename,
+                        borehole["borehole_index"],
+                        layer_index,
+                        language,
+                        layer["material_description"],
+                        uscs_class,
+                        None,
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/utils/data_loader.py:73]:
```suggestion
                        prediction_uscs_class=None,
```

File: src/description_classification/utils/data_loader.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                    )
+                )
+
+    return layer_descriptions
+
+
+def get_data_language_count(layer_descriptions: list[LayerInformations]) -> dict[str:int]:
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/utils/data_loader.py:80]: `dict[str:int]` not correct type?

File: src/description_classification/utils/data_loader.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        layer_descriptions (list[LayerInformations]): All the layers.
+
+    Returns:
+        dict[str:int]: the count for each language.
+    """
+    language_counts = dict(Counter(layer.language for layer in layer_descriptions))
+    return language_counts
+
+
+def get_data_class_count(layer_descriptions: list[LayerInformations]) -> dict[str:int]:
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/utils/data_loader.py:93]: `dict[str:int]` not correct type?

File: src/description_classification/utils/data_loader.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        layer_descriptions (list[LayerInformations]): All the layers.
+
+    Returns:
+        dict[str:int]: the count for each class.
+    """
+    class_counts = dict(Counter(layer.ground_truth_uscs_class.name for layer in layer_descriptions))
+    return class_counts
+
+
+def write_predictions(layers_with_predictions: list[LayerInformations], out_dir: Path):
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/utils/data_loader.py:106]: writing predictions does not really relate to "data loader", maybe put this in a separate python file?

File: src/description_classification/evaluation/evaluate.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    Attributes:
+        global_metrics (dict[USCSClasses, Metrics]): A dictionary containing the classification metrics
+            for each of the USCS classes at a global level.
+        language_metrics (dict[str, dict[USCSClasses, Metrics]]): A dictionary where each key represents
+            a supported language. Each value is another dictionary containing the classification metrics
+            for each of the USCS classes in that language.
+    """
+
+    global_metrics: dict[USCSClasses:Metrics]
+    language_metrics: dict[str : dict[USCSClasses:Metrics]]
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/evaluation/evaluate.py:34]: Dict type with , instead of :

File: src/description_classification/evaluation/evaluate.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        precisions = []
+        recalls = []
+        f1s = []
+
+        for m in metric_list:
+            tp, fp, fn = m.tp, m.fp, m.fn
+
+            precision = tp / (tp + fp) if (tp + fp) > 0 else 0.0
+            recall = tp / (tp + fn) if (tp + fn) > 0 else 0.0
+            f1 = 2 * precision * recall / (precision + recall) if (precision + recall) > 0 else 0.0
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/evaluation/evaluate.py:61]: Why not use the precision, recall and f1 methods from the Metrics class?

File: src/description_classification/evaluation/evaluate.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        """
+        if not metric_list:
+            return {"micro_precision": 0, "micro_recall": 0, "micro_f1": 0}
+
+        all_aggregated_metric = Metrics.micro_average(metric_list)
+
+        return {
+            "micro_precision": round(all_aggregated_metric.precision, 2),
+            "micro_recall": round(all_aggregated_metric.recall, 2),
+            "micro_f1": round(all_aggregated_metric.f1, 2),
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/evaluation/evaluate.py:96]: 4 decimal places instead of 2

File: src/description_classification/evaluation/evaluate.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+    return all_classification_metrics
+
+
+def per_class_metrics(predictions: list[USCSClasses], ground_truth: list[USCSClasses]) -> dict[USCSClasses, Metrics]:
+    """Compute per-class classification metrics.
+
+    Args:
+        predictions (List[str]): List of predicted class labels.
+        ground_truth (List[str]): List of actual class labels.
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/evaluation/evaluate.py:233]: types not consistent

Comment (stijnvermeeren-swisstopo) [reply src/description_classification/evaluation/evaluate.py:233]: Suggestion to avoid the zip:
```
def per_class_metrics(layers: list[LayerInformations]) -> dict[USCSClasses, Metrics]:
    all_classes = set([layer.prediction_uscs_class for layer in layers]) | set([layer.ground_truth_uscs_class for layer in layers])
    metrics_per_class = {}

    for cls in all_classes:
        tp = sum(1 for layer in layers if layer.prediction_uscs_class == layer.ground_truth_uscs_class == cls)
...
```

File: src/description_classification/evaluation/evaluate.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+
+def log_metrics_to_mlflow(all_classification_metrics: AllClassificationMetrics, log_per_class=False):
+    """Log metrics to MLFlow with error handling."""
+    # Log overall metrics
+    for name, value in all_classification_metrics.to_json().items():
+        mlflow.log_metric(name, value)
+
+    # Log per-class metrics if requested
+    if log_per_class:
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/evaluation/evaluate.py:258]: Consensus to totally remove this flag and always log the per-class metrics.

File: src/description_classification/evaluation/evaluate.py [hunk 12]
```diff
@@ -192,69 +182,63 @@ def to_json_per_class(self) -> dict[str, float]:
...
 
 
-def evaluate(layer_descriptions: list[LayerInformations], log_per_class: bool = False) -> AllClassificationMetrics:
+def evaluate(layer_descriptions: list[LayerInformations]) -> AllClassificationMetrics:
     """Evaluates the predictions of the LayerInformations objects against the ground truth.
 
     Args:
-        layer_descriptions (list[LayerDescriptionWithGroundTruth]): the LayerInformations objects
+        layer_descriptions (list[LayerInformations]): the LayerInformations objects
         log_per_class (bool, optional): whether to log per-class metrics. Defaults to False.
```
Comment (stijnvermeeren-swisstopo) [inline src/description_classification/evaluation/evaluate.py:190]: arg should be removed from doc string

**Review by stijnvermeeren-swisstopo — CHANGES_REQUESTED** [review]
one more small thing to clean up

**Review by stijnvermeeren-swisstopo — APPROVED** [review]
